### ISSUE TITLE
OPFS asyncer: in waitLoop(): Error: No waitLoop handler for whichOp # 0

### ISSUE THREAD

**Issue by VegarRingdalAibel** [sqlite/sqlite-wasm#12]
Might be something Im doing wrong.
Atm it feels more random if it works or not, and I think its because I have 5 workers/5 databases

![image](https://github.com/tomayac/sqlite-wasm/assets/94840334/23b43754-4be8-4bd2-9bf3-710af2a2e69a)

![image](https://github.com/tomayac/sqlite-wasm/assets/94840334/697fa62b-435a-4f58-9c91-688541d8594d)

![image](https://github.com/tomayac/sqlite-wasm/assets/94840334/5f7117a1-bacb-493d-8870-67596a87243d)

Code running
![image](https://github.com/tomayac/sqlite-wasm/assets/94840334/44687f6d-d6bf-419d-af95-b5dea26b5b0c)

Also tried this one `console.log(this.db.affirmOpen()?.pointer, this.db.affirmOpen()?.filename)`

It it weird all 5 show same pointer? 
I do not know enough to say if this is to be expected....
![image](https://github.com/tomayac/sqlite-wasm/assets/94840334/10738aca-a752-42bb-8334-a4be10035c32)

![image](https://github.com/tomayac/sqlite-wasm/assets/94840334/a7efd229-c16f-4bba-8927-3e962ca2f8cf)

**Comment by tomayac** [comment]
@VegarRingdalAibel As [per the `README`](https://github.com/tomayac/sqlite-wasm#sqlite-wasm), I'm closing this Issue here. @sgbeal, this might be a worthwhile problem to look into on your side, though.

**Comment by VegarRingdalAibel** [comment]
@tomayac 
sorry missed that one. My bad.

@sgbeal tell me if you want me to create issue under sqllite / or you create a issue. Or you know what Im doing wrong.

**Comment by sgbeal** [comment]
> It it weird all 5 show same pointer?

Not weird at all: each worker has its own independent virtual machine and memory address space. If all 5 start up identically, it's unsurprising that they have the same virtual addresses.

As to the error you're seeing, i have no clue as to the cause. It's never come up in 9ish months of testing. Where it's happening is clear, but why it happens is currently inconceivable to me. It seems very possible that the number of workers is the problem. Frankly, i consider 5 to be expecting too much from a browser environment, given the high memory requirements of workers and limited file-sharing capabilities of OPFS. Some degree of moderation is required.

(please pardon brevity - currently living on a boat and limited to a phone for internet access.)

**Comment by sgbeal** [comment]
> Btw, only seen this issue when Ive created indexes or delete a lot.

That is an interesting hint. A few weeks ago we fixed a JS-side sync-related bug for journal files which could _hypothetically_ be a factor in this problem. IIRC that fix went into 3.42.x but i will dig that out the next time i have access to an AC power outlet (in the next couple of days).

In the meantime, if you can manage to reproduce it reliably and post a tree which does it (public or direct to me, as you prefer: stephan at wanderinghorse.net), that would be a tremendous help.

> Every worker have own file/db, so its not really sharing the file with anyone else. But yes it might be a bug in OPFS.

In that case there "shouldn't" be any contention issues with OPFS directly, but it is _hypothetically_ feasible that that number of high-I/O workers is causing the browser to hiccup. Can you try it in Firefox v111 or higher, to hopefully rule out (or in!) a browser-specific quirk?

**Comment by VegarRingdalAibel** [comment]
Ill make a solution during the week so its easy for you guys to just clone/run to see the issue.

btw, did a little more testing now.
The issue goes away if I never do delete/recreate index in all worker at once, just do those in sequence (let one worker finish before asking the next one to do the same)

So its like every worker is fighting for resources and maybe timeout to quickly?
I dont know the limits of OPFS...
But Ill try and see if same issue is in firefox later too.

**Comment by sgbeal** [comment]
After some testing, i've gone ahead and made the above change, squelching (i believe) that particular instance of that particular console error. There _appears_ to be a _potential_ window for an exception to be thrown which, in turn, leads to the side effect of the "No waitLoop handler for whichOp # 0", in which case that report is harmless but unduly scary. That change will be in v3.43. The [prerelease snapshot](https://sqlite.org/wasm/uv/snapshot.html) has been updated with this change.

@VegarRingdalAibel please be sure to try the "unlock-asap" mode to see if that helps as well:

<https://sqlite.org/wasm/doc/trunk/persistence.md#opfs-unlock-asap>

i don't believe it will for your particular setup, but it can't hurt to try. It should have noticeable effects on concurrent access to the same db file, though.

**Comment by VegarRingdalAibel** [comment]
👍 
Ill give a try later today.

I was thinking if its getting timeout/fighting for resources we might need to up the asyncIdleWaitTime/maxTries ?
Is that a option I can pass inn to it?

**Comment by vegarringdal** [comment]
Made a small weird sample here
https://github.com/vegarringdal/sqllitedebug

Not tried to improve anything, needed something that broke first..
Important to have dev tools closed, will break faster.


![msedge_gwdy8NB1Ca](https://github.com/tomayac/sqlite-wasm/assets/2901416/9bce1e6b-9904-4f7b-bbfd-a91346af3c85)

**Comment by Matt-TOTW** [comment]
Hello, I'm experiencing this issue, or one very similar to it.
Like vegarringdalI am also running multiple workers, where each worker is connecting to its own db/opfs file so I don't see how concurrency could be an issue. At no point are separate threads connecting to the same db. Here are the console warnings and errors 
![image](https://github.com/sqlite/sqlite-wasm/assets/36234834/efca60e7-e9c8-4da2-bd48-77bf32cdcaab)
It's probably worth mentioning that I often don't see that first error
![image](https://github.com/sqlite/sqlite-wasm/assets/36234834/85e54fae-54c7-4940-a234-f11372c3a587)

When I slow things down so that the connections don't occur simultaneously, all is well. Also, it seems much more likely to happen with longer running queries.
Each db is quite large ~1.5Gb.

There is one difference that I am seeing compared to vegarringdalI. When I revert the suggested changes ie back to
`if('timed-out'===Atomics.wait(`
I am not seeing the `No waitLoop handler for whichOp` error - so I'm not certain where exactly the problem is happening.

@vegarringdal, did the above solve your problem?

**Comment by sgbeal** [comment]
> When I slow things down so that the connections don't occur simultaneously, all is well. Also, it seems much more likely to happen with longer running queries.

This is the first time in more than a year the problem has been reported and we unfortunately know nothing about it beyond the other comments in this thread. In short, it appears to be browser-specific, only happens under specific severe loads, and manifests as an "impossible" result code.

If you do not require any concurrency, i'd recommend trying the other OPFS VFS, which does not require the Atomics API and is much faster (the main tradeoff being a complete lack of concurrency):

https://sqlite.org/wasm/doc/trunk/persistence.md#vfs-opfs-sahpool

**Comment by Matt-TOTW** [comment]
Thanks @sgbeal, a bit of a shame but that does seem fair. I confess I hadn't given the sahpool much attention as I do eventually want to support multi-tabbing, but having read Hashimoto's suggestions for client-level solutions it seems well worth trying. 
Thanks!

**Comment by vegarringdal** [comment]
Did try much more since it was to unstable for what I tried to do. And a was not ready for special headers needed for sharedarray. Got those in place last month.
Sahpool sounds interesting, want to look more into that one 👍

**Comment by Matt-TOTW** [comment]
Sorry to keep banging on, but is this anything? I will be the first to admit that I don't fully understand the code. All I've done is introduce a waitLoop in the `opRun` function of `thePromise` to ensure that the `rc` value really has changed away from -1.

![image](https://github.com/sqlite/sqlite-wasm/assets/36234834/5ce2f461-2d98-47a4-9227-f790dd57ea87)
Thinking about it, I'm now wondering if this disables some legitimate scenarios.

Somewhere somehow, the code in the async-proxy is presumably notifying without changing `rc` away from -1. Again, I don't really understand the code.

This has solved the problem insofar as I am not seeing those errors. What I haven't yet been able to verify yet is if the operation is complete and correct.

What do you think? Am I barking up the wrong tree?

**Comment by sgbeal** [comment]
There's only one code path which changes, and notifies about, the rc, and my eyes have never been able to spot somewhere where the rc does not get set to one of the SQLITE_xyz values (none of which are negative). Of course, a bug cannot be ruled out, but i've never been able to spot it. Barring the existence of such a bug, if the Atomics API is behaving such that that loop is required, it hints at a browser bug, in that it violates (unless i'm sorely mistaken) the Atomics API.

**Comment by Matt-TOTW** [comment]
It took me a while but I've finally confirmed that, as you surmised, this is browser (Chrome) specific - well at least it does not happen on Firefox.
So I suppose the thing to do would be to submit a bug report to Chrome? If I do that I will try and create a minimal repro but am I right that it's something like...

Under certain heavy-load conditions we see the following:

Reading thread:
`const waitResult = Atomics.wait(typedArray, index, x)`

Writing thread:
`Atomics.store(typedArray, index, y)`
`Atomics.notify(typedArray, index)`

Reading thread:
`waitResult` is `ok`

In other words, on the reading thread, the value y is not obtained.

How did I do?

**Comment by sgbeal** [comment]
That sounds like a fair summary to me, with the note that `x` and `y` are different values in that case.

Here's a relevant snippet from the current upstream code:

```
        if('not-equal'!==Atomics.wait(
          state.sabOPView, state.opIds.whichOp, 0, state.asyncIdleWaitTime
        )){
          /* Maintenance note: we compare against 'not-equal' because

             https://github.com/tomayac/sqlite-wasm/issues/12

             is reporting that this occassionally, under high loads,
             returns 'ok', which leads to the whichOp being 0 (which
             isn't a valid operation ID and leads to an exception,
             along with a corresponding ugly console log
             message). Unfortunately, the conditions for that cannot
             be reliably reproduced. The only place in our code which
             writes a 0 to the state.opIds.whichOp SharedArrayBuffer
             index is a few lines down from here, and that instance
             is required in order for clear communication between
             the sync half of this proxy and this half.
          */
          await releaseImplicitLocks();
          continue;
        }
...
```

**Comment by Matt-TOTW** [comment]
So far at least, I have not been able to create a minimal repro for this. I might give it another go over the weekend.

I think I slightly misunderstood `Atomics.wait`. The problem we are seeing is not that it resolves to `ok`, it's simply that `y` is not found in the array by the reading thread even though it was set in the array by the writing thread.
I thought that `wait()` resolves with `not-equal` if the value is different _on waking_, but actually `wait()` will only return `not-equal` when the value doesn't match the test value when `wait()` is _called_.

**Comment by Matt-TOTW** [comment]
I think I have reproduced it and if I have done so correctly then I am seeing it on all browsers and not just Chrome.
https://github.com/Matt-TOTW/simultaneous-multithread-atomics

Despite the name of the repo, I sometimes see the error without the need to work simultaneously, the `for` loop is enough to trigger it.
I'd be interested to know what you think, is this even a fair reproduction?
Thanks!

**Comment by sgbeal** [comment]
i've never seen the problem myself so i hesitate to say whether that's a fair or unfair repro :/. My _vague (mis?)understanding_ of it is that it's dependent on high throughput, rather than the number of threads, but that's really just guesswork on my part. My understanding of the Atomics API is that it "should not be possible" for `Atomics.wait()` to return `'ok'` in the cases where it is doing so.

That said: it is absolutely possible that there's a bug here on our end, but repeated attempts to spot it have not discovered it.

**Comment by Matt-TOTW** [comment]
> My understanding of the Atomics API is that it "should not be possible" for Atomics.wait() to return 'ok' in the cases where it is doing so.

I'm not sure I agree. My understanding is that `ok` is always the expected result when awoken by a `notify`, irrespective of the value.
- `timed-out` is returned and program resumes after this number of ms without being awoken by a `notify`
- `ok` is returned and program resumes when it is awoken by a `notify`, no matter what the value is
- `not-equal` is returned _immediately_ (the program is never halted, ie a `notify` will have no effect) if the value currently at the given index is not the given value.

The problem as I see it is that `wait` is being awoken but the value loaded at the slot in question is not what was stored there by the other thread, even though it was stored by the other thread before it emitted the notification.

**Comment by sgbeal** [comment]
> I'm not sure I agree. My understanding is that...

You are absolutely correct - i was working from a broken recollection and typing from a tablet so didn't go double-check myself.

> The problem as I see it is that wait is being awoken but the value loaded at the slot in question is not what was stored there by the other thread, even though it was stored by the other thread before it emitted the notification.

Agreed - that's my current understanding.

**Comment by Matt-TOTW** [comment]
So is this ?
1. A problem with how browsers implement the Atomics API
2. A simple bug somewhere in sqlite-wasm
3. Or is sqlite-wasm somehow misusing the atomics API

I think it is arguably 3. I think that the waitLoop and the way the two threads interact isn't properly thread-safe. I've tried to diagram out a scenario where a problem could occur but it got quite complicated and so I think it's simpler to show it in code. (Happy to share the diagram though if you think it'd help). The below will always produce the error.
![image](https://github.com/sqlite/sqlite-wasm/assets/36234834/17cfe411-e782-4601-86b9-81e4156f0f2c)
[branch](https://github.com/Matt-TOTW/simultaneous-multithread-atomics/tree/Add-pauses)
Maybe it's a bit weird dropping in long pauses here and there like that, but who knows what goes on at the low level at which these APIs operate, especially when under heavy load, right? The main point is the two threads are fully isolated with no guarantees of sequence of interthread operations, so I think something like this is possible. When I see this error occur in my real project it is generally part of a read, and I can see that the `opRun` function is called ~500,000 times for a single read.

Would you agree that this might suggest a problem?
The only way I can think to solve it is to introduce the `checkLoop` like I did in [my earlier comment](https://github.com/sqlite/sqlite-wasm/issues/12#issuecomment-2210480058) which essentially ensures the first thread waits for the second one if that one falls a loop behind. (I think... I'd want to double check that).

**Comment by Matt-TOTW** [comment]
Thanks sgbeal, 
> The Atomics.wait/notify coordination, in our case, always involves precisely two threads, no more, no less. One thread wait()s on the other to notify(), at which point wait() returns control to its caller. Perhaps i'm being completely naive, but it seems pretty difficult to misuse that, nor do i see any other ways of making that thread-safer than the guarantees provided by the wait()/notify() APIs.

What I think I have proven is that the threads can get out of sync, essentially leading to an errant notify() (from a slow running previous call/iteration) kicking a wait() out of sleep too early (and before the value has been set).

> i would argue that any behavior which doesn't match the API docs is either a doc bug or a browser-side bug, and the behavior we're seeing here does not match my reading of MDN's docs. It's also possible, of course, that MDN's docs simply don't cover certain cases.

I'm not saying there's any non-conformant behaviour, it seems to me that this behaviour can emerge just from how the two threads wait(), store() and notify() each other when there's no guarantee of execution order between the threads.
For example, consider:
```
Atomics.wait(typedArray, index, -1) === 'not-equal';
const result = Atomics.load(typedArray, index);
```
We'd expect `result` not to equal `-1` but there is no guarantee of this as the other thread way have snuck in a different value between those two statements. (Although to be clear, I'm not saying that's exactly what's happened in our case). This doesn't break the atomicity of those two statements, it's just the implementation doesn't allow for that scenario.

Thanks for looking into it, and sorry if I'm misunderstanding. Hope it cools down!

**Comment by sgbeal** [comment]
> What I think I have proven is that the threads can get out of sync, essentially leading to an errant notify()

And what i believe that proves is that there's a browser bug ;).

> We'd expect result not to equal -1 but there is no guarantee of this as the other thread may have snuck in a different value after the wait() and before the load().

Whereas i would very much _not_ expect that because only two threads are running and they are 100% coordinated by the Atomics.wait/notify ordering. When one thread is wait()ing, the other is running, and vice versa. It should _never_ be the case that the second thread is running unless the first thread wait()ing. i don't see how the second thread (the "async proxy," in this case) can be doing _any_ work unless the first thread is wait()ing on it.

**Comment by Matt-TOTW** [comment]
If you run the code in my [branch](https://github.com/Matt-TOTW/simultaneous-multithread-atomics/tree/Add-pauses), do you see the error? Maybe my system is being weird (although I did confirm it on chrome and Firefox).

The if wait() then load() example I gave is very contrived and not what I think is happening, I was just trying to illustrate a point about Atomics and two threads. What I think might be happening is illustrated in that branch. I don't believe it's demonstrative of the Atomics api misbehaving - unless you think those pauses are so contrived as to be practically impossible?
Sorry to keep banging on. I do agree that the wait/notify coordination is thread-safe, it's just that when you throw in store()s and load()s as well I believe that might not be.

If you're not seeing the error in the branch, I'll shut up. I promise. 🙂

I've been reading about mutexes... I wonder if they might help. 

https://blogtitle.github.io/using-javascript-sharedarraybuffers-and-atomics/

**Comment by sgbeal** [comment]
i haven't yet looked at your demo closely - and won't until our temperatures drop back down to the mid 20s (supposedly tomorrow) - but there's no path in the VFS's async proxy which writes to the SAB after the proxy has called notify(). The proxy only accesses the SAB when the first half of the VFS is wait()ing on it. There's no reason (per the Atomics API docs, anyway) why anything should appear randomly in the SAB.

As far as mutexes go: Atomics.wait is the mutex, for all practical purposes. We lock it, do work on the other side, and release it with notify(). When wait() is waiting, only the async half of the VFS may access the SAB, and when wait() is not held then only the other half is accessing it. Once notify() is called from the async half, that half does nothing except sit in its `waitLoop()` waiting for the next wait() call.  The linked mutex impl isn't doing anything we don't already do except adding complexity (e.g. support for multiple clients of the mutex, which we don't need because the VFS halves always operate in matched pairs). For a more general-purpose problem that's a very interesting class, though.

If we had multiple workers sharing the same instance of the async proxy, i would have to assume that the problem is a race condition bug on our side, but we always have matched pairs of two and they share work in a 100% linear fashion, which rules out a race condition. i just triple-checked, and the only operation the async proxy performs after calling notify() is updating its own internal debugging metrics (which have no effect on its behavior). After calling notify(), then updating the metrics, it always returns immediately to its wait-loop to wait on the next wait() call.

**Comment by Matt-TOTW** [comment]
Does this make sense? I've created a Miro board that you should be able to zoom around on - hopefully it'll also work on a tablet. I've tried to step through the sqlite-wasm code on each thread to try and explain what I think might be happening.

Nb That repro I've mentioned I think faithfully represents the mechanics of the interplay between the two threads, and the branch I've pointed to manually enforces those pauses, but hopefully this Miro will also show you what I mean.
Please let me know what you think, thanks!
https://miro.com/app/board/uXjVK0HRpq4=/?share_link_id=982565601824

Here is an image of it as well in case that's easier.
![image](https://github.com/sqlite/sqlite-wasm/assets/36234834/89c176ff-90b5-496f-99df-765da7fdfe95)

Ps I'll happily take some of those high temperatures, summer has yet to arrive in the UK :(

**Comment by sgbeal** [comment]
Wow, thank you for that extremely detailed analysis! The temperatures have finally dropped to the point where sitting on the computer is not torturous.

i'm still trying to wrap my head around (A) _how_ the "errant op" case can happen and (B) how we can recover from that. Is it correct to say that if thread 1 calls `Atomics.notify()` and thread 2 isn't actually sitting at `Atomics.wait()` at that instant, then we can get into this unfortunate situation?

Regarding your [wait change proposal](#issuecomment-2210480058): i don't have a better idea and do not see any fundamental flaw in doing that. It makes me a bit nervous because i'm not confident that i fully understand how this hiccup is happening, and, as you say:

> What I haven't yet been able to verify yet is if the operation is complete and correct.

On the other hand...

After going through your analysis at length, but before looking closely at that change proposal, my instinct was to propose a change which was fundamentally the same as yours, so it seems that i'm finally catching up to your line of thinking.

So... i'll go ahead and apply an equivalent patch of:

```js
      while( 'not-equals' !== Atomics.wait(state.sabOPView, state.opIds.rc, -1) ){
        /* See discussion at https://github.com/sqlite/sqlite-wasm/issues/12 */
      }
```

and see where that leads us.

My test suite is not nearly as "stressful" as your case, so if you would please (re)try this change in your own tree(s), that would improve my confidence about it!

**Comment by sgbeal** [comment]
BTW:

> If you run the code in my [branch](https://github.com/Matt-TOTW/simultaneous-multithread-atomics/tree/Add-pauses), do you see the error?

Yes, on both FF and Chrome on Linux.

**Comment by Matt-TOTW** [comment]
Excellent, thanks!!

>  Is it correct to say that if thread 1 calls `Atomics.notify()` and thread 2 isn't actually sitting at `Atomics.wait()` at that instant, then we can get into this unfortunate situation?

Yes, I think so (excluding the path with the `continue`).  The first key thing that I think happens is that thread 2 gets ahead and `store()`s the value before thread 1 even calls `wait()`. Then, thread 1 cycles back round (with another call to opRun) and gets to `wait` before thread 2 `notify()`s.

Here's something else, `Atomics.notify()` returns the count of the number of agents that were waiting at that index, so it occurred to me that just before I see the error, the `notify` from thread 1 should be `0` as per the logic I have outlined. But to my surprise I see `0` a lot, and too often to meaningly correlate with the error. Does this surprise you?

I will definitely try your suggestion and let you know asap - apologies in advance if it's not today.

**Comment by sgbeal** [comment]
> The first key thing that I think happens is that thread 2 gets ahead and store()s the value before thread 1 even calls wait().

That's definitely not inconceivable. Hmm.

> Here's something else, Atomics.notify() returns the count of the number of agents that were waiting at that index, so it occurred to me that just before I see the error, the notify from thread 1 should be 0 as per the logic I have outlined. But to my surprise I see 0 a lot, and too often to meaningly correlate with the error. Does this surprise you?

That is exceptionally curious. That notify() might actually return something interesting had never occurred to me :/.

> I will definitely try your suggestion and let you know asap - apologies in advance if it's not today.

No worries and no hurry. Thank you tons for your assistance, persistence, and detective work on this!

For completeness's sake, this change is currently in a branch:

<https://sqlite.org/src/timeline?r=opfs-errant-op>

The plan is to merge that into trunk as soon as we're reasonably satisfied that it doesn't cause any new grief. So far, so good.

**Comment by VegarRingdalAibel** [comment]
Amazing work guys 🥇 

That `performance.now()`, can manage to create issues, slow it down and create issues ?

![image](https://github.com/sqlite/sqlite-wasm/assets/94840334/2717389f-0edc-4278-b36f-40239209dea6)

**Comment by sgbeal** [comment]
> That performance.now(), can ... slow it down and create issues ?

An excellent question. That cannot be ruled out and gives me some motivation for removing the timing/metrics pieces altogether. We used that data extensively during the initial development so that we could optimize better and provide some timing data to interested browser developers, but we have not used those metrics in almost two years, so it's really just wasted CPU cycles now.

**Comment by sgbeal** [comment]
The metrics-tracking bits have been removed from [this ticket's feature branch](https://sqlite.org/src/timeline?r=opfs-errant-op).

**Comment by vegarringdal** [comment]
That was quick 👍 
Is there any temp builds we can try out with this change (removed metrics), to see if it makes a difference ?

**Comment by vegarringdal** [comment]
Updated https://github.com/vegarringdal/sqllitedebug to 3.46.0-build2

![image](https://github.com/sqlite/sqlite-wasm/assets/2901416/5166700d-c067-4ac0-a5e9-c40b03b24f03)

Since i dont get this issue when dev tools is open, at least not that fast it would have been nice to just try it out.

**Comment by sgbeal** [comment]
> Is there any temp builds we can try out with this change (removed metrics), to see if it makes a difference ?

Here's a short-lived link to a build of this branch:

https://wasm-testing.sqlite.org/tmp/sqlite-wasm-snapshot-20240711-3470000.zip

> Updated https://github.com/vegarringdal/sqllitedebug to 3.46.0-build2

That error is from the new build? Does it also show up in the old build?

> Since i dont get this issue when dev tools is open, at least not that fast it would have been nice to just try it out.

Having the dev tools open changes invisible things which can indeed affect how the VFS performs. In my experience the VFS is really slow if the dev tools are open, even if no console output is being generated.

**Comment by vegarringdal** [comment]
> > Is there any temp builds we can try out with this change (removed metrics), to see if it makes a difference ?
> 
> Here's a short-lived link to a build of this branch:
> 
> https://wasm-testing.sqlite.org/tmp/sqlite-wasm-snapshot-20240711-3470000.zip
> 
> > Updated https://github.com/vegarringdal/sqllitedebug to 3.46.0-build2
> 
> That error is from the new build? Does it also show up in the old build?
> 
> > Since i dont get this issue when dev tools is open, at least not that fast it would have been nice to just try it out.
> 
> Having the dev tools open changes invisible things which can indeed affect how the VFS performs. In my experience the VFS is really slow if the dev tools are open, even if no console output is being generated.

@sgbeal 
Looks like my issues is gone... usually never reach batch2 before its failed, now its been running for minutes without issues.

![image](https://github.com/sqlite/sqlite-wasm/assets/2901416/f01bc857-ea6d-4888-ae36-4363f12778ec)


@Matt-TOTW Maybe you can give your sample a try with this version too ?

**Comment by vegarringdal** [comment]
@Matt-TOTW 
Amazing work with the Miro board btw 🥇

**Comment by sgbeal** [comment]
> Looks like my issues is gone...

Great! That's encouraging!

> ... now its been running for minutes without issues.

But is it behaving as you'd expect? No mysteriously missing data or some such? 🤔

**Comment by vegarringdal** [comment]
I never had missing data, it just broke down on me when I was using many db at once/doing a lot of work in workers.
Hopefully all issues is solved, will be fun to see if it solves @Matt-TOTW  issues.

Dont have any way to run my old project anymore.
Im using sqllite still, but without going crazy with to many db open at once to avoid prev errors. But maybe I can go crazy again.😂

**Comment by Matt-TOTW** [comment]
Hey guys, sorry for the radio silence.
I've put that while loop you suggested sgbeal in place
```
while( 'not-equal' !== Atomics.wait(state.sabOPView, state.opIds.rc, -1) ){
  /* See discussion at https://github.com/sqlite/sqlite-wasm/issues/12 */
}
```
and I am not receiving the error and as far as I can tell, the data looks good.

I also tried doing this on the notify in `opRun`
```
while (Atomics.notify(state.sabOPView, state.opIds.whichOp) < 1);
```
ie
![image](https://github.com/user-attachments/assets/3206fa9d-3c4e-4b72-bd57-873fa472f5b4)
which also works.

Nb, when I did that on the `storeAndNotify()` `notify()`, it hung. So I guess that function is utilised in more than one pathway and it's probably nothing to worry about.

I personally quite like this second approach as it speaks more directly to the problem (we think).

**Comment by Matt-TOTW** [comment]
Ps I didn't notice an appreciable drop in the error rate without the `performance.now()` call. Just fyi. But that's a _very_ back-of-the-envelope assessment.

**Comment by Matt-TOTW** [comment]
I found this discussion interesting. 

https://github.com/tc39/ecma262/issues/1492

empijey asserts that "the right way to deterministically use wait is looping"

So perhaps that is the better approach after all.

**Comment by sgbeal** [comment]
From that discussion:

> they must normally be wrapped in a little bit of protocol or they will be unreliable (lost wakeups).

"Lost wakeup" might be a good description for this thread's problem, but i like the "errant op" description from the Miro board.

As to which loop to go with, i'm ambivalent. i agree that the notify()<1 speaks better to what we believe the problem is, but i'm also not 100% confident that i understand the problem ;). One advantage the 'not-equal' loop has is that we know that it works for @vegarringdal (this ticket's OP) ;). The questions about the notify() result code approach don't have a clear final answer in the above-linked issue, either.

My current preference, but not a strong one, would be to go with the 'not-equals' loop, but i am definitely willing to switch (or even do both?) if we're convinced that it/they would be better.

**Comment by sgbeal** [comment]
This fix is now in the upstream trunk and the 3.46 branch. It's as-yet unknown whether there will be another 3.46.x patch release, but if there is then this fix will be included.

Thank you both for your support on this, especially @Matt-TOTW for your persistence and detective work!

**Comment by Matt-TOTW** [comment]
Great! You're very welcome. It's a very powerful project, thanks for maintaining!
